cranelift/isle/isle/isle_examples/link/multi_constructor_main.rs

```diff
@@ -0,0 +1,42 @@
+mod multi_constructor;
+
+pub(crate) type ConstructorVec<T> = Vec<T>;
+
+struct Context;
+
+struct It {
+    i: u32,
+    limit: u32,
+}
+
+impl multi_constructor::ContextIter for It {
+    type Context = Context;
+    type Output = u32;
+    fn next(&mut self, _ctx: &mut Self::Context) -> Option<u32> {
+        if self.i >= self.limit {
+            None
+        } else {
+            let i = self.i;
+            self.i += 1;
+            Some(i)
+        }
+    }
+}
+
+impl multi_constructor::Context for Context {
+    type etor_C_iter = It;
+    fn etor_C(&mut self, value: u32) -> Option<It> {
+        Some(It { i: 0, limit: value })
+    }
+
+    fn ctor_B(&mut self, value: u32) -> Option<Vec<u32>> {
+        Some((0..value).rev().collect())
+    }
```

Comment:
I would love for that to work out, but the problem here is ownership: there is not necessarily any other owner for a newly-generated set of return values from a multi-constructor. Take for example a top-level "simplify" entry point that returns multiple equivalent expressions for a given input expression: it will need to return a vector of expression/node/eclass IDs, but this vec doesn't exist anywhere else (and stashing it in a side-table just to return an iterator over it would be awkward because it doesn't need to persist beyond the postprocessing we do on it).

Reply:
I haven't read the surrounding code, but in general, an `Iterator` can itself own the data over which it's iterating. In particular, `Vec::into_iter` returns an iterator that owns the vector and will drop it when the iterator is dropped.

Reply:
Ah, that's a good point! I'll rework in this direction.

Reply:
Updated just now. It's a little awkward with the `ContextIter` adapters (this is necessary to allow the lowering trait to provide multi-etor iterators without holding a borrow on the whole `Context`) but it seems to work.